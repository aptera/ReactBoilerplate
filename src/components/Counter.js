import React, { Component } from 'react';
import '../styles/counter.scss';

class Counter extends Component {
    constructor(props) {
        super(props);
        this.state = { counter: 0 };
    }

    // https://facebook.github.io/react/docs/react-component.html#componentdidmount
    componentDidMount() {
        this.interval = setInterval(this.tick.bind(this), 1000);
    }

    tick() {
        this.setState({
            counter: this.state.counter + 1
        });
    }

    // https://facebook.github.io/react/docs/react-component.html#componentwillunmount
    componentWillUnmount() {
        clearInterval(this.interval);
    }

    render() {
        return (
            <h2>{this.state.counter} seconds since the last refresh</h2>
        );
    }
}

export default Counter;