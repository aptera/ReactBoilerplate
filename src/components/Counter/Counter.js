import React, { Component } from 'react';
import './counter.scss';

class Counter extends Component {
    constructor(props) {
        super(props);
        this.state = { counter: 0 };
    }

    // https://facebook.github.io/react/docs/react-component.html#componentdidmount
    componentDidMount() {
        setTimeout(this.tick.bind(this), 1000);
    }

    tick() {
        this.setState({
            counter: this.state.counter + 1
        });
        setTimeout(this.tick.bind(this), 1000);
    }

    render() {
        return (
            <h2>{this.state.counter} seconds since the last refresh</h2>
        );
    }
}

export default Counter;