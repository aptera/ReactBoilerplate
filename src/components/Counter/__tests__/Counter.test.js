import React from 'react';

import renderer from 'react-test-renderer';
import ReactTestUtils from 'react-dom/test-utils';

// import the component to put under test.
import Counter from '../Counter';

var CounterComponent;
jest.useFakeTimers();

describe('the counter', () => {

    beforeEach(() => {
        CounterComponent = ReactTestUtils.renderIntoDocument(<Counter />);
    });

    it('renders correctly', () => {
        const tree = renderer.create(
            <Counter />
        ).toJSON();
        expect(tree).toMatchSnapshot();
    });

    // Using the beforeEach above, interact with the components' functions.
    it('increments the counter for each tick', () => {
        CounterComponent.tick();
        expect(CounterComponent.state.counter).toBe(1);
        CounterComponent.tick();
        expect(CounterComponent.state.counter).toBe(2);
        CounterComponent.tick();
        expect(CounterComponent.state.counter).toBe(3);
        CounterComponent.tick();
        CounterComponent.tick();
        expect(CounterComponent.state.counter).toBe(5);
    });

    it('calls the tick function each second', () => {

        // This is important, because other tests will impact values.  Especially if using a beforeEach.
        jest.resetAllMocks();

        // To ensure the above worked, this expectation should pass.
        expect(setTimeout.mock.calls.length).toBe(0);

        CounterComponent = ReactTestUtils.renderIntoDocument(<Counter />);
        CounterComponent.tick = jest.fn();

        expect(setTimeout.mock.calls.length).toBe(1);
        expect(setTimeout.mock.calls[0][1]).toBe(1000);

        // tick 5 times
        CounterComponent.tick();
        CounterComponent.tick();
        CounterComponent.tick();
        CounterComponent.tick();
        CounterComponent.tick();        
        
        // tick() should have been called 5 times.
        expect(CounterComponent.tick).toHaveBeenCalledTimes(5);

        // how do you test an effing timer? :(
    });
});