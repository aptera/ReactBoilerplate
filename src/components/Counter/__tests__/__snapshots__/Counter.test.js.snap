// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the counter renders correctly 1`] = `
<h2>
  0
   seconds since the last refresh
</h2>
`;
